File: backend/app/routers/stats.py
"""Stats API."""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from .. import stats
from ..database import get_db
from ..deps import require_auth
from ..models import Server

router = APIRouter(prefix="/stats", tags=["stats"])


def _server(db: Session, server_id: int) -> Server:
    server = db.get(Server, server_id)
    if server is None:
        raise HTTPException(status_code=404, detail="服务器不存在")
    return server


def _metrics(metric: str = "", metrics: list[str] | None = None) -> list[str]:
    values = list(metrics or [])
    if metric:
        values.append(metric)
    if not values:
        values.append("custom.play_time")
    return values


@router.get("/metrics")
def metrics(
    q: str = Query(""),
    category: str = Query("all", pattern="^(all|important|normal|ignored)$"),
    limit: int = Query(200, ge=1, le=1000),
    _: object = Depends(require_auth),
    db: Session = Depends(get_db),
) -> list[dict]:
    return stats.list_metrics(db, q=q, category=category, limit=limit)


@router.get("/{server_id}/players")
def players(
    server_id: int,
    limit: int = Query(200, ge=1, le=1000),
    _: object = Depends(require_auth),
    db: Session = Depends(get_db),
) -> dict:
    _server(db, server_id)
    return {"players": stats.player_options(db, server_id, limit)}


@router.get("/{server_id}/leaderboard")
def leaderboard(
    server_id: int,
    metric: str = Query(""),
    metrics: list[str] = Query(default=[]),
    window: str = Query("total"),
    limit: int = Query(100, ge=1, le=1000),
    _: object = Depends(require_auth),
    db: Session = Depends(get_db),
) -> dict:
    _server(db, server_id)
    metric_keys = _metrics(metric, metrics)
    return {
        "rows": stats.leaderboard(db, server_id, metric_keys, window, limit),
        "scanned_at": stats.scanned_at(server_id),
    }


@router.post("/{server_id}/refresh")
async def refresh(
    server_id: int,
    scope: str = Query("important", pattern="^(important|full)$"),
    _: object = Depends(require_auth),
    db: Session = Depends(get_db),
) -> dict:
    server = _server(db, server_id)
    await stats.scan_server_async(server, scope)
    return {"scanned_at": stats.scanned_at(server_id, scope)}


@router.get("/{server_id}/series")
def series(
    server_id: int,
    uuid: str = Query(""),
    uuids: list[str] = Query(default=[]),
    metric: str = Query(""),
    metrics: list[str] = Query(default=[]),
    mode: str = Query("delta", pattern="^(delta|total)$"),
    granularity: str = Query("10min"),
    hours: int = Query(24, ge=1, le=8760),
    _: object = Depends(require_auth),
    db: Session = Depends(get_db),
) -> dict:
    _server(db, server_id)
    ids = list(uuids)
    if uuid:
        ids.append(uuid)
    return {"points": stats.series(db, server_id, ids, _metrics(metric, metrics), mode, granularity, hours)}

File: web/src/api/stats.ts
import { apiRequest } from '@/api/client'

export type StatSampleType = 'important' | 'normal' | 'ignored'

export interface StatMetric {
  key: string
  label: string
  sample_type: StatSampleType
}

export interface StatPlayer {
  uuid: string
  name: string
}

export interface LeaderRow {
  uuid: string
  name: string
  value: number
}

function paramsOf(input: Record<string, string | number | string[] | undefined>) {
  const p = new URLSearchParams()
  Object.entries(input).forEach(([key, value]) => {
    if (value === undefined) return
    if (Array.isArray(value)) value.forEach((v) => p.append(key, v))
    else p.set(key, String(value))
  })
  return p.toString()
}

export function listStatMetrics(category = 'all', q = '', limit = 300): Promise<StatMetric[]> {
  return apiRequest<StatMetric[]>(`/stats/metrics?${paramsOf({ category, q, limit })}`)
}

export function listStatPlayers(serverId: number): Promise<{ players: StatPlayer[] }> {
  return apiRequest(`/stats/${serverId}/players`)
}

export function getLeaderboard(
  serverId: number,
  metrics: string[],
  window: string,
  limit = 100,
): Promise<{ rows: LeaderRow[]; scanned_at: number | null }> {
  return apiRequest(`/stats/${serverId}/leaderboard?${paramsOf({ metrics, window, limit })}`)
}

export function refreshStats(serverId: number, scope: 'important' | 'full' = 'important'): Promise<{ scanned_at: number | null }> {
  return apiRequest(`/stats/${serverId}/refresh?${paramsOf({ scope })}`, { method: 'POST', body: '{}' })
}

export function getStatSeries(args: {
  serverId: number
  uuids: string[]
  metrics: string[]
  mode: 'delta' | 'total'
  granularity: string
  hours: number
}): Promise<{ points: Record<string, [number, number][]> }> {
  const { serverId, ...rest } = args
  return apiRequest(`/stats/${serverId}/series?${paramsOf(rest)}`)
}
